subprojects/platform-base/src/main/java/org/gradle/language/base/plugins/LifecycleBasePlugin.java

```diff
@@ -59,14 +62,33 @@ public void apply(final ProjectInternal project) {
     }
 
     private void addClean(final ProjectInternal project) {
-        addPlaceholderAction(project, CLEAN_TASK_NAME, Delete.class, new Action<Delete>() {
+        final Callable<File> buildDir = new Callable<File>() {
+            public File call() throws Exception {
+                return project.getBuildDir();
+            }
+        };
+
+        addPlaceholderAction(project, CLEAN_TASK_NAME, Clean.class, new Action<Delete>() {
             @Override
             public void execute(Delete clean) {
                 clean.setDescription("Deletes the build directory.");
                 clean.setGroup(BUILD_GROUP);
-                clean.delete(new Callable<File>() {
-                    public File call() throws Exception {
-                        return project.getBuildDir();
+                clean.delete(buildDir);
+            }
+        });
+
+        // Register at least the project buildDir as a directory to be deleted.
+        final BuildOutputCleanupRegistry buildOutputCleanupRegistry = project.getServices().get(BuildOutputCleanupRegistry.class);
+        buildOutputCleanupRegistry.registerOutputs(buildDir);
+
+        // Register custom targets to be deleted, if configured.
+        project.getTasks().withType(Clean.class, new Action<Clean>() {
```

Comment:
IIUC, the only reason for the `Clean` task is to be able to match it here. Can we instead add these actions in the block after line 74 instead? Or can we match by task name instead?